### PR TITLE
Poc refactor

### PR DESCRIPTION
**Pull request by marcin-zyrkowski**
This is the whole Remote Terminal functionality according to SoW.
Plus a little more
  - a separate config file mender-connect.conf   (for compatibility with Go client)
  - an config option to define a shell while default is /bin/sh
  - the configuration reads the 'user' entry but this is not used - agreed that this is out of scope

### CODE REVIEW

File: src/mender/remoteterminal/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 Northern.tech AS
```
Comment (pasinskim) [inline src/mender/remoteterminal/__init__.py]: Should be 2021 :)

File: src/mender/remoteterminal/remoteterminal.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
+# Copyright 2020 Northern.tech AS
```
Comment (pasinskim) [inline src/mender/remoteterminal/remoteterminal.py]: Same here: 2021

File: setup.py [hunk 3]
```diff
@@ -44,10 +44,10 @@
     ],
     keywords=["mender", "OTA", "updater"],
     packages=setuptools.find_packages(where="src"),
-    install_requires=["cryptography", "requests"],
+    install_requires=["cryptography", "requests", "msgpack", "websockets"],
     entry_points={"console_scripts": ["mender-python-client=mender.mender:main"]},
     package_dir={"": "src"},
-    python_requires=">=3.6",
+    python_requires=">=3.9",
```
Comment (pasinskim) [inline setup.py]: Is this still a requirement?

File: src/mender/remoteterminal/remoteterminal.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                packed_msg = await self.client.recv()
+                msg: dict = msgpack.unpackb(packed_msg, raw=False)
+                hdr = msg['hdr']
+                if hdr['typ'] == MESSAGE_TYPE_SPAWN_SHELL and not self.session_started:
+                    self.sid = hdr['sid']
+                    self.session_started = True
+                    await self.send_client_status_to_backend(MESSAGE_TYPE_SPAWN_SHELL)
+                    self.open_terminal()
+                    self.start_transmitting_thread()
+                if hdr['typ'] == MESSAGE_TYPE_SHELL_COMMAND:
```
Comment (alfrunes) [inline src/mender/remoteterminal/remoteterminal.py]:
```suggestion
                elif hdr['typ'] == MESSAGE_TYPE_SHELL_COMMAND:
```

File: src/mender/remoteterminal/remoteterminal.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                    self.write_command_to_shell(msg)
+                if hdr['typ'] == MESSAGE_TYPE_STOP_SHELL:
```
Comment (alfrunes) [inline src/mender/remoteterminal/remoteterminal.py]:
```suggestion
                elif hdr['typ'] == MESSAGE_TYPE_STOP_SHELL:
```

File: src/mender/remoteterminal/remoteterminal.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                    self.shell.kill()
+                    self.master = None
+                    self.slave = None
+                    await self.send_client_status_to_backend(MESSAGE_TYPE_STOP_SHELL)
+                    self.session_started = False
```
Comment (alfrunes) [inline src/mender/remoteterminal/remoteterminal.py:101]: (Not required) Does not handle "resize" message types.

**Review by alfrunes — CHANGES_REQUESTED** [review]
This implementation does not respect the protocol multiplexing of multiple sessions in parallel. If you open up multiple terminals in parallel you will end up with zombie shell processes. IOW, the open terminal processes should be indexed by the `sid` header in the incoming packages. The open processes also needs to be monitored in a separate thread so that the main thread knows if a terminal dies (e.g. by `exit`) and can notify the other peer with a "stop" message.

File: src/mender/remoteterminal/__init__.py (hunk 1 above)
Comment (marcin-zyrkowski) [reply src/mender/remoteterminal/__init__.py]: Will be fixed.

File: src/mender/remoteterminal/remoteterminal.py (hunk 2 above)
Comment (marcin-zyrkowski) [reply src/mender/remoteterminal/remoteterminal.py]: Will be fixed.

File: setup.py (hunk 3 above)
Comment (marcin-zyrkowski) [reply setup.py]: If we don't support a 'user' option, and we don't, then this is not a requirement anymore.
Will be reverted to 3.6

File: src/mender/remoteterminal/remoteterminal.py (hunk 6 above)
Comment (marcin-zyrkowski) [reply src/mender/remoteterminal/remoteterminal.py:101]: According to the SoW we don't support resizing in this version.

**Pull request merged**